전남대 BE_서영우 2주차 과제 (3단계)

westzeroright opened the pull request.

westzeroright committed:
chore: 상품 관리 코드 옮겨 오기

westzeroright committed:
style(ProductDao): sql문 코드 포맷팅

westzeroright committed:
style(ProductController): 메서드 사이 공백 추가

메서드 사이에 공백을 넣어 가독성을 향상시켰습니다.

westzeroright committed:
chore(build.gradle): validation 의존성 추가

westzeroright committed:
docs(README): 1단계 구현 기능 목록 작성

westzeroright committed:
feat(ProductRequestDto,ProductController): 유효성 검증

상품 이름에 관하여 요구사항을 지키도록 유효성 검증 실행

westzeroright committed:
feat(GloablExceptionHandler): 예외처리

유효성 검사 오류에 대하여 처리

westzeroright committed:
feat(Member): 회원 도메인

이메일과 비밀번호를 가진다

westzeroright committed:
feat(MemberController): 회원 추가 기능

http 메서드는 POST이며, url은 /member/register이다. 도메인 모델을 직접 사용하며 반환한다.

westzeroright committed:
chore(build.gradle): JWT 의존성 추가

westzeroright committed:
feat(MemberDao): 회원 가입시 회원 정보 DB에 저장

westzeroright committed:
feat(JwtTokenUtil): JWT 토큰 발급

westzeroright committed:
feat(schema.sql): create member table

westzeroright committed:
feat(TokenResponseDto): 토큰을 json으로 반환하기 위한 객체

westzeroright committed:
feat(MemberController): 회원가입 기능

이메일과 비밀번호를 DB에 저장하고 access token 발급

westzeroright committed:
feat(MemberDao): 회원 조회

westzeroright committed:
feat(MemberController): 회원 로그인 기능

이메일과 비밀번호로 로그인을 시도하면 DB에 일치하는 회원이 있는지 확인 후 access token 발급

westzeroright committed:
refactor(MemberController): 중복 회원가입 방지

이미 가입한 회원에 대해 예외처리

westzeroright committed:
docs(README): 2단계 구현 기능 목록 작성

westzeroright committed:
chore: 패키지 설정

westzeroright committed:
feat(LoginMember): LoginMember 어노테이션 만들기

westzeroright committed:
feat(LoginMemberArgumentResolver): HandlerMethodArgumentResolver를 구현한 클래스

두 메서드 구현
- supportsParameter (해당 메서드의 매개변수가 해당 resolver를 지원하는지 확인)
- resolveArgument (request 헤더에서 토큰을 꺼내고 토큰을 통해 얻은 Member 객체를 반환)

westzeroright committed:
feat(WebCong): 구현한 HandelrMethodArgumentResolver 구현

westzeroright committed:
feat(schema.sql): create table wishlist

product 테이블과 member 테이블을 참조

westzeroright committed:
refactor(schema.sql): memberId 사용자 설정으로 변경

westzeroright committed:
feat(WishRequestDto): 장바구니 기능 요청 클래스

westzeroright committed:
docs(README): 3단계 구현 기능 목록 작성

westzeroright committed:
feat(JwtTokenUtil): JWT 토큰 디코딩

westzeroright committed:
refactor(LoginMemberArgumentResolver): 토큰으로부터 회원 아이디 얻기

westzeroright committed:
feat(MemberDao): 회원 아이디로 회원 조회

westzeroright committed:
feat(WishlistController): 장바구니 CRUD

westzeroright committed:
feat(WishlistDao): 장바구니 테이블 접근

westzeroright committed:
style(JwtTokenUtil): 코드 포맷팅

subject를 email로 명시

westzeroright committed:
fix(LoginMemberArgumentResolver): 어노테이션 위치는 파라미터

westzeroright committed:
refactor(MemberDao): 아이디에서 이메일로 토큰 발급 수정

로그인&회원가입시 이메일로 토큰 발급

westzeroright committed:
refactor(WishlistController): 장바구니 기능 url 매핑

westzeroright committed:
refactor(Wishlist): wishProduct id만 반환

wishlist 테이블에 해당 상품이 있는지 확ㅇ니

westzeroright committed:
Merge branch 'westzeroright' into step3

westzeroright committed:
fix(schema.sql): wishlist 테이블 수정

기본키 제거

westzeroright committed:
delete: step2 브랜치 이동과정에서 문제

westzeroright committed:
chore: 패키지 구조 수정

pkeugine commented on README.md:
```diff
@@ -14,3 +14,21 @@
+  - password
+
+  
+email과 password를 서버로 요청을 보내고 DBD에 일치한 회원이 있는지 확인 후 access token 발급
```
오타가 있는 것 같아요.
그리고 문서를 조금 더 구체적으로 작성하고, 개발하며 잘 활용해보면 어떨까요?

예를 들어 개발을 시작하기 전에 todo list 처럼 만들어야 할 기능을 나열하고,
commit 을 하면서 하나씩 체크하는거죠.

```markdown
- [x] 이렇게 체크 표시를 사용해볼 수도 있어요
- [ ] 이러면 완료되지 않은 작업을 나타내는거구요
```

westzeroright replied:
넵 알겠습니다!

pkeugine commented on build.gradle:
```diff
@@ -25,6 +25,9 @@ dependencies {
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
     implementation 'org.springframework.boot:spring-boot-starter-validation'
+    compileOnly 'io.jsonwebtoken:jjwt-api:0.12.6'
+    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.12.6'
+    runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.12.6'
```
다른 의존성과는 다르게 이건 버전이 명시되어 있어요.
다른 코드나 글을 참고해서 넣어주셨겠지만, 이 참에 어떤 차이가 있는지 확인해보면 좋을 것 같아요.

westzeroright replied:
스프링 부트 Gradle plugin은 스프링 부트 의존성 관리 및 애플리케이션 패키이징을 지원하고 
io.spring.dependency-management 플러그인을 적용하면 사용중인 스프링 부트 버전에 맞는 의존성을 자동으로 가져올 수 있는 것을 확인했습니다. 특별히 버전을 명시해서 사용하는 경우 해당 버전으로 오버라이딩되어 사용된다고 확인하였습니다.

pkeugine commented on src/main/java/gift/login/JwtTokenUtil.java:
```diff
@@ -0,0 +1,35 @@
+
+    public Claims decodeJWT(String token) {
+            return Jwts.parser().setSigningKey(key).build().parseClaimsJws(token).getBody();
+    }
```
위의 방식과는 다르게 `decodeJWT` 는 한 줄에 메서드 체이닝을 해주셨네요. 방식을 비슷하게 유지해보죠!

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/login/JwtTokenUtil.java:
```diff
@@ -0,0 +1,35 @@
+            .compact();
+    }
+
+    public Claims decodeJWT(String token) {
```
`decodeJWT` 라는 이름을 대중적인 컨벤션에 맞게 수정해보면 좋을 것 같아요.
[google java style guide 라는 컨벤션을](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case) 굉장히 많은 개발자분들이 참고하고 적용하는 편인데,
이 내용을 한 번 확인해보죠.

뭔가 한 번 봤던 리뷰같지 않나요? :)
[제가 한 번 남겼던 적이 있는 내용](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-product/pull/145#discussion_r1659608078)이네요.
아무래도 리뷰가 짧은 시간동안 많이 생기다보니 놓친 부분이 있을 것 같아 다시 남겨보아요!

westzeroright replied:
으악 다시 한 번 정리하겠습니다..

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/login/LoginMember.java:
```diff
@@ -0,0 +1,48 @@
+import org.springframework.web.method.support.HandlerMethodArgumentResolver;
+import org.springframework.web.method.support.ModelAndViewContainer;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+public @interface LoginMember {
```
커스텀 어노테이션!
잘 만들어주셨네요. 👍 👍 👍 
`@Retention` 과 `@Target` 은 각각 어떤 내용을 나타내는건가요?
그리고 여기에 들어가는 값들은 `RUNTIME` 과 `PARAMETER` 외에 어떤 것들이 있는지,
각각 무엇을 의미하는지도 알아보죠!

westzeroright replied:
`@Retention` : 해당 어노테이션을 언제까지 유지할 것인지를 정의하는 부분

- 종류
    - `@Retention(RetentionPolicy.SOURCE)` - 컴파일 이후에도 계속 참조 가능
    - `@Retention(RetentionPolicy.CLASS)`  - 컴파일러가 클래스를 참조할 때까지 유표
    - `@Retention(RetentionPolicy.RUNTIME)` - 컴파일 전까지만 유효

`@Target` : Java compiler가 annotation이 어디에 적용될지 결정하기 위해 사용

```
ElementType.PACKAGE : 패키지 선언
ElementType.TYPE : 타입 선언
ElementType.ANNOTATION_TYPE : 어노테이션 타입 선언
ElementType.CONSTRUCTOR : 생성자 선언
ElementType.FIELD : 멤버 변수 선언
ElementType.LOCAL_VARIABLE : 지역 변수 선언
ElementType.METHOD : 메서드 선언
ElementType.PARAMETER : 전달인자 선언
ElementType.TYPE_PARAMETER : 전달인자 타입 선언
ElementType.TYPE_USE : 타입 선언
```
이렇게 알아보았습니다!

사실 과제 진행 중에 `@Target(ElementType.PARAMETER)` 이 부분을 메서드로 잘못 지정하여 막혔었는데 이번에 공부하게 되어서 좋네요:)

pkeugine commented on src/main/java/gift/login/LoginMember.java:
```diff
@@ -0,0 +1,48 @@
+            NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
+
+            String token = webRequest.getHeader("Authorization").substring(7);
+            System.out.println("토큰"+token);
```
값을 출력해서 console 창에서 확인할 수 있도록 해주셨군요 👍 
`System.out.println` 말고 로깅을 적용해보시면 어떨까요?

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/login/LoginMember.java:
```diff
@@ -0,0 +1,48 @@
+public @interface LoginMember {
+
+    @Component
+    class LoginMemberArgumentResolver implements HandlerMethodArgumentResolver {
```
해당 클래스는 어노테이션 안에 두기보단 별도의 자체 파일로 관리해봐도 좋을 것 같아요.

westzeroright replied:
원래 따로 클래스 파일이 있었으나 step2 브랜치와 step3 브랜치를 오가며 이런 문제가 발생한 것 같습니다🥲

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/member/Member.java:
```diff
@@ -0,0 +1,62 @@
+    private String name;
+    private String email;
+    private String password;
+    private boolean role;
```
충분히 필드로 알 수 있는 내용인 것 같으니 주석은 제거해보죠!

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/member/MemberController.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    @GetMapping
+    public void getMember(@LoginMember Member member) {
+        System.out.println(member.getEmail()+member.getPassword());
+    }
```
이 메서드는 어떤 용도로 만드셨을까요?

westzeroright replied:
제거했습니다..!

pkeugine commented on src/main/java/gift/product/GlobalExceptionHandler.java:
```diff
@@ -1,29 +1,15 @@
-import org.springframework.web.bind.annotation.RestController;
 import org.springframework.web.bind.annotation.RestControllerAdvice;
 
-//@RestControllerAdvice
-//public class GlobalExceptionHandler {
-//
-//    @ResponseStatus(value = HttpStatus.BAD_REQUEST)
-//    @ExceptionHandler(MethodArgumentNotValidException.class)
-//    public String handlePrdouctNameException(MethodArgumentNotValidException e) {
-//        return e.getBindingResult().getAllErrors().getFirst().getDefaultMessage();
-//    }
-//}
-
```
주석 제거 리뷰 반영 👍

pkeugine commented on src/main/java/gift/wishlist/WishlistController.java:
```diff
@@ -0,0 +1,57 @@
+
+    @DeleteMapping("/{id}")
+    public HttpEntity<String> deleteWish(@PathVariable(name="id") Long wishId) {
+//        if (wishlistDao.findProductById(wishId).isEmpty()) {
+//            throw new NoSuchElementException("잘못된 접근입니다");
+//        }
+//        else {
+//            wishlistDao.deleteWish(wishId);
+//        }
```
주석을 제거하자는 리뷰를 이번 pr 에 반영해주셔서 기분이 좋았는데...
다른 부분에서 생겨서 마음이 아픕니다 ㅠㅜ

westzeroright replied:
ㅜㅜ저도요..😢

pkeugine commented on src/main/java/gift/wishlist/WishlistController.java:
```diff
@@ -0,0 +1,57 @@
+//        else {
+//            wishlistDao.deleteWish(wishId);
+//        }
+        if (wishlistDao.findProductById(wishId).isEmpty()) {
+            throw new NoSuchElementException("잘못된 접근입니다");
+        }
```
그래도 예외를 던지고, GlobalExceptionHandler 에서 이를 처리해주는 것은 좋네요 :)

pkeugine commented on src/main/java/gift/wishlist/WishlistDao.java:
```diff
@@ -0,0 +1,61 @@
+@Repository
+public class WishlistDao {
+
+    private final ProductDao productDao;
```
해당 필드는 어떤 이유로 추가해주신걸까요?
일단 사용하지 않으므로 제거해주죠!

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/wishlist/WishlistDao.java:
```diff
@@ -0,0 +1,61 @@
+    private final ProductDao productDao;
+    private JdbcClient jdbcClient;
+
+    public WishlistDao(JdbcClient jdbcClient, ProductDao productDao) {this.jdbcClient = jdbcClient;
+        this.productDao = productDao;
+    }
```
코드가 이렇게 되는건 지양해주세요!

westzeroright replied:
반영했습니다!

pkeugine commented on src/main/java/gift/wishlist/WishlistController.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/wishes")
```
wish 와 wishlist 라는 표현이 함께 사용되고 있는데, 이를 하나로 통일해보면 좋을 것 같아요.

westzeroright replied:
반영했습니다..!

pkeugine commented on src/main/java/gift/member/Member.java:
```diff
@@ -0,0 +1,62 @@
+        this.role = role;
+    }
+
+    public Member() {
+
+    }
```
이 생성자는 어떠한 이유로 추가되었나요?

westzeroright replied:
엇 분명 필요해서 추가했는데 지워도 오류가 나지 않은 것 같으니 필요가 없는 것 같습니다

pkeugine reviewed (changes requested):
안녕하세요 영우님
리뷰 남겨보았으니 한 번 확인해주시고, 질문에 대한 답변은 댓글로 남겨주세요!

이번 step 에 이전 리뷰들 중 반영되지 않은 것들도 챙겨주시면 좋을 것 같아요.
파이팅입니다! 🔥 🔥 🔥

pkeugine commented:
코멘트가 하나 누락됐네요! 로그아웃 기능도 만들어보도록 하죠 :)

westzeroright committed:
add: 삭제된 파일 복구

westzeroright committed:
add: 삭제된 파일 복구

westzeroright committed:
docs: 리드미 오타 수정

westzeroright committed:
style: 메서드 체이닝 통일

westzeroright committed:
style: 메서드 컨벤션 수정

westzeroright committed:
style: 메서드 컨벤션 수정

westzeroright committed:
refactor: 사용하지 않는 필드 제거

westzeroright committed:
style: 불필요한 주석 제거

westzeroright committed:
style: 표현 통일

westzeroright committed:
refactor: 불필요한 생성자 제거

westzeroright committed:
style: 안 쓰는 import문 제거

westzeroright committed:
chore: 로깅을 위한 룸북 의존성 추가

westzeroright committed:
log: token, email 로깅

westzeroright committed:
style: 불필요한 주석 제거

westzeroright committed:
fix: select 오류 해결

westzeroright committed:
refator: 클래스 위치 수정 변경

westzeroright committed:
delete: 불필요한 메서드 제거

westzeroright committed:
refactor: JwtTokenUtil을 객체로 만들지 않음

- 빈으로 등록하지 않기
- MemberController에 주입해주는 부분 제거
JwtTokenUtil은 유틸리티 클래스이기 때문에 bean으로 관리할 필요가 없다

westzeroright committed:
refactor: 토큰 만료 시간 magic number로 설정

westzeroright committed:
refactor(Member): 사용하지 않는 setter 제거

westzeroright committed:
refacor: url 리소스 표기 변경 - api/복수형

westzeroright committed:
style: SQL 대문자 변경 및 인덴트 활용

westzeroright committed:
fix: 버그 수정(bean 설정)

westzeroright committed:
refactor: `ResponseEntity.badRequest()` 활용

westzeroright committed:
refactor: price, url 검증 메세지 출력

westzeroright committed:
refactor: kakao, Kakao, KAKAO 등과 같은 표현 역시 검증

westzeroright committed:
test(ProductControllerTest): POST 테스트 코드

westzeroright committed:
test(ProductControllerTest): 모든 상품을 조회하는 GET 테스트 코드

westzeroright committed:
style: SQL 대문자 변경 및 인덴트 적용

westzeroright committed:
test: 아이디로 상품 조회 GET 테스트 코드

westzeroright committed:
test: 아이디로 상품 삭제 DELETE 테스트 코드

westzeroright committed:
test: 아이디로 상품 수정 UPDATE 테스트 코드

westzeroright committed:
refactor: 테스트 코드 중복된 부분 @BeforeEach 설정

westzeroright committed:
feat: 로그아웃시 토큰 저장 블랙리스트 테이블

westzeroright committed:
feat: 토큰 관리 DAO

로그아웃시 블랙리스트 테이블에 토큰을 저장하고 로그인시 토큰이 존재하는지 찾습니다

westzeroright committed:
feat: 로그아웃 기능

인증 헤더로부터 토큰을 받아 블랙리스트 테이블에 저장

westzeroright committed:
feat: deprecated 된 시크릿키 설정 변경

westzeroright commented:
> 코멘트가 하나 누락됐네요! 로그아웃 기능도 만들어보도록 하죠 :)

반영했습니다!

westzeroright commented:
멘토님 1주차, 2주차 피드백 모두 3주차 0단계 PR에 걸쳐 모두 반영했습니다:)

pkeugine reviewed (approved):
다음 미션 진행중이라 이만 머지하도록 하겠습니다.
리뷰 확인하시느라 고생하셨어요 👍

pkeugine commented:
conflict 가 발생하고 있네요.
이 부분 해결해주시면 머지하도록 하겠습니다.

westzeroright committed:
refactor: conflict 해결용

westzeroright committed:
refactor: conflict 해결용
